### DISCUSSION TITLE
Error when trying to setup the unit test environment in the Math Compoenent Tutorial

### DISCUSSION THREAD

**Discussion by NishqR** [nasa/fprime#1226] [Q&A]
Hi,

I'm doing the Math Component tutorial with the latest version, and have followed the whole process pretty thoroughly - but have been stuck on section 4.5 with the following error - 

![image](https://user-images.githubusercontent.com/49609432/149711081-0cd857ed-b0b6-4424-9a89-79d029abd66a.png)

I've looked around for this error but I think some of the solutions are not pertinent to the latest version. I've also tried starting the tutorial from scratch thrice just to ensure I didn't make any mistake as I went along. 

Just wondering if anyone could help me out with this. I'm running the latest version of ubuntu on virtualbox if that helps!

**Comment by LeStarch** [comment]
This is a known issue with our tools: https://github.com/nasa/fprime/issues/1195.  It has been fixed, and we are working on the next release.

**Reply by NishqR** [reply]
Awesome, thank you for the reply! Is there an ETA on the next release?

**Reply by LeStarch** [reply]
I released the tools yesterday.  You can upgrade with the following:

```
pip install -U fprime-tools
```

**Reply by NishqR** [reply]
I updated the tools, did a fresh clone and tried but it gave me a new error - 

![image](https://user-images.githubusercontent.com/49609432/150516265-98158976-b763-44ce-891c-a6a81eb6aaab.png)

So i ran `fprime-util generate --ut`, which ran fine but after that I got the same error again. I then tried the workaround you mentioned in [#1195](https://github.com/nasa/fprime/issues/1195), but I'm just getting this: 

![image](https://user-images.githubusercontent.com/49609432/150516665-292fb75e-f6c9-4f73-9d16-924c2952b7ce.png)

Not entirely sure if what Ive done is correct, could you please let me know if I've made any errors with the workaround? 

Thanks!

**Comment by abdelrahman47** [comment]
[https://github.com/fprime-community/fprime-tools/pull/35/files/72317b96d880ddc267615921d8e02a314c174464
](https://github.com/fprime-community/fprime-tools/pull/35/files/72317b96d880ddc267615921d8e02a314c174464)

Check this solution. It worked for me

**Reply by NishqR** [reply]
Hey thanks for the reply, I checked this file builder.py and the change seems to be already in there, must've updated with fprime-tools. Would you know any other reason why this is persisting?

**Reply by astroesteban** [reply]
I'm seeing the same error on my end too. I got the latest devel branch of F' running in an Ubuntu 20.04 Docker image.

I installed the latest prime-tools from source and I still see the error.

I'll spend some time debugging today and report back if I come across a fix for this.

**Reply by LeStarch** [reply]
Can you post the version of fprime-tools and fprime?

```
pip freeze | grep fprime-tools
```

and

```
cd <frime>
git describe --tags --always
```

**Reply by LeStarch** [reply]
I confirmed it is working as it is working on the latest `devel` of both. Thus we need to track down the versions.

**Reply by NishqR** [reply]
Cheers @LeStarch, here's the output from both: 

![image](https://user-images.githubusercontent.com/49609432/151285158-ad767469-2b6f-49bb-8c11-2c6fdb0fdebf.png)

Did a fresh clone this morning and can confirm the error is still persisting unfortunately

**Reply by LeStarch** [reply]
I was able to confirm that is working on my side with those two versions.  Probably should have checked these items earlier:

1. Are you in a component directory when running `fprime-util impl --ut`
2. Have you added `register_fprime_ut()` to your `CMakeLists.txt`
3. Have you tried removing and regenerating your build cache
```
cd Ref
rm -r build-fprime-automatic-native-ut
fprime-util generate --ut
cd <componet dir>
fprime-util impl --ut
```

**Reply by LeStarch** [reply]
I ask because these errors can (with working tools) occur for the three reasons above. From the error posted here, item 1 is confirmed.

**Reply by NishqR** [reply]
Thanks for checking @LeStarch! I'm doing the math component tutorial and can confirm the following:

1. Running this in the MathSender component folder
2. I tried adding this in both the `CMakeLists.txt` in `Ref/` and also `MathSender`, and it threw the following error when I tried to run `fprime-util generate --ut` in Ref:
![image](https://user-images.githubusercontent.com/49609432/151475761-6a5980ae-2865-4e9f-878d-76996bc2e913.png) 
Tried running `fprime-util purge` as well as `rm -r build-fprime-automatic-native-ut`, but it kept throwing the same error. Just wondering if there's more I need to add, since in the Math Component tutorial this is a later step - https://github.com/nasa/fprime/blob/devel/docs/Tutorials/MathComponent/Tutorial.md#45-write-and-run-unit-tests. I'm currently unable to even generate the Tester files

3. As mentioned in point 2, tried this with and without adding `register_fprime_ut()` in `CMakeLists.txt`, and I'm still getting the same error when going into the `MathSender` folder and running `fprime-util impl --ut`:
![image](https://user-images.githubusercontent.com/49609432/151477185-a4b3766c-79f2-471c-b045-3b18044d2ec7.png)


Do you think there's perhaps a step missing in the MathComponent tutorial itself?

**Reply by astroesteban** [reply]
So I was getting the same error. I fixed it by:

1. Adding the following to `Ref/MathSender/CMakeLists.txt`:
```cmake
Register the unit test build
set(UT_SOURCE_FILES
  "${CMAKE_CURRENT_LIST_DIR}/MathSender.fpp"
  "${CMAKE_CURRENT_LIST_DIR}/test/ut/Tester.cpp"
  "${CMAKE_CURRENT_LIST_DIR}/test/ut/main.cpp"
)
set(UT_MOD_DEPS STest)
register_fprime_ut()
```

2. Then I made the test folders: `mkdir -p test/ut`

3. Then I made two placeholder test files:
```sh
touch test/ut/Tester.cpp
touch test/ut/main.cpp
```

4. Ran the `fprime-util generate --ut`

5. Finally I ran `fprime-util impl --ut` and copied the newly generated files into the `test/ut` folder.

That got everything working.

**Reply by NishqR** [reply]
Oh man cheers for that that worked! Can you confirm if that didn't cause any issues with the rest of the tutorial? Think if not that step definitely needs to come before we run `fprime-util impl --ut` for the first time

**Reply by astroesteban** [reply]
Great! I went through the rest of the test tutorial section and everything worked. Looks like the docs just needed some steps rearranged.

**Reply by NishqR** [reply]
That's great to hear, I'll complete it properly when I have the chance but thanks for the help, appreciate it! :)

**Reply by LeStarch** [reply]
@astroesteban thanks for the pointers here.  I am going to take an action on the tools to make this process less frustrating.   Several fixes we can do:

1. Translating missing `testimpl` to "did you forget register_fprime_ut()" style error
2. Ensure unit test files generate in the expected `test/ut` folder.  

Anything else I should capture here?

**Reply by astroesteban** [reply]
Great! I think you captured everything @LeStarch.
FYI: I opened a pull request (#1243) that just has the documentation fixes I mentioned above.
This discussion also correlates with issue #1238.
